docs/design/secure-config.md

```diff
@@ -0,0 +1,228 @@
+5. The victim unzips the repo at an arbitrary location, make some changes, then run `jj fix`
+6. They have now executed an arbitrary command on the victim’s system
+
+This attack vector can be solved by ensuring that we can determine the path that the repo was stored at.
+
+#### Attack vector 3: Replay attack with some social engineering to preserve paths
+
+1. The victim uploads a zip file of a repository they have locally on their system at `/path/to/repo`
+1. The attacker can now see any cryptographic signatures stored in the repository
+2. The attacker runs `jj config set --repo fix.tools.foo ‘[“malicious”, “command”]’`
+3. The attacker copies the victim’s cryptographic signature and puts it in their malicious repository.
+4. The attacker zips up their repo, sends it to the victim, and instructs them to install it at `/path/to/repo`
+5. The victim unzips the repo at `/path/to/repo`, make some changes, then run `jj fix`
+6. They have now executed an arbitrary command on the victim’s system
+
+This attack vector can be solved by making repository configuration untamperable.
+
+#### Attack vector 4: Extremely advanced replay attack with insecure code
+
+1. The victim creates a repo
+2. The victim runs `jj config set --repo fix.tools.foo = [“$repo/format.py”]`
+3. The victim uploads a zip file of a repository they have locally on their system at `/path/to/repo`
+1. The attacker can now see any cryptographic signatures stored in the repository
+4. The attacker modifies `format.py` to be malicious
+5. The attacker zips up their repo and sends it to the victim
+6. The victim runs `jj fix`
+7. They have now executed an arbitrary command on the victim’s system
+
+This attack vector cannot feasibly be dealt with. It would require a signature of the transitive closure of files that can be accessed via jj configs to solve.
+
+## Objective 
+
+### Goals
+
+* Prevent as many of the above attack vectors as possible
+* Have minimal negative impacts on UX
+
+### Non-goals (Optional)
+
+* Use strategies such as sandboxing to mitigate damage
+  * We could do this for formatters, for example, but then repo hooks would have the same problem
+  * These options are not mutually exclusive
+
+## Detailed Design
+
+### Storing config out-of-repo
+
+We will store per-repo configuration in `etcetera::BaseStrategy::config_dir().join(“jj”).join(“repos”).join(id)`. 
+
+The structure of a repo will look like so:
+
+```
+$HOME/.config/jj/
+  repos/
+    abc123/
+      metadata.binpb
+      config.toml
+  workspaces/
+    def456/
+      config.toml
+      metadata.binpb
+my-repo/.jj/
+  workspace-id (contains "def456")
+  workspace-config.toml (symlink to $HOME/.config/jj/workspaces/def456/config.toml)
+  repo/
+    repo-id (contains "abc123")
+    config.toml (symlink to $HOME/.config/jj/repos/abc123/config.toml)
+```
+Metadata.binpb will refer to the following protobuf:
+
+```proto
+message Metadata {
+  // This is used to distinguish between copies and moves.
```

Comment:
This is very unclear to me (at least at this point in the doc). I don't even know what "distinguish between copies and moves" means. Is it `cp -r ~/src/jj ~/src/jj2` vs `mv ~/src/jj ~/jj2`? Why do we want to distinguish between them? I would suggest either adding more detail here or removing the comment so the reader is not left confused.